gemini/use-cases/retrieval-augmented-generation/qna_using_query_routing/utils/qna_vector_search.py

```diff
@@ -0,0 +1,144 @@
+        3. If the answer to the question cannot be determined from the context alone, say "I cannot determine the answer to that."
+        4. If the context is empty, just say "I could not find any references that are directly related to your question."
+
+        Context:
+        =============
+        {context}
+        =============
+
+        What is the Detailed explanation of the answer to the following question?
+        Question: {question}
+        Detailed explanation of Answer:"""  # pylint: disable=line-too-long
+
+    def find_relevant_context(self, query: str) -> str:
+        """
+        Searches the vector index to retrieve relevant context based on a query embedding.
+
+        Args:
+            query (str, optional): The query text.
+
+        Returns:
+            str: The concatenated text of relevant documents found in the index.
+        """
+
+        # Generate the embeddings for user question
+        vector = self.text_embedding_model.get_embeddings([query])
+
+        queries = [vector[0].values]
+
+        response = self.index_endpoint.find_neighbors(
+            deployed_index_id=self.deployed_index_id,
+            queries=queries,
+            num_neighbors=self.num_neighbors,
+        )
+
+        context = ""
+        for neighbor_index in range(len(response[0])):
+            context = (
+                context
+                + self.embedding_df[
+                    (self.embedding_df["id"] == response[0][neighbor_index].id)
+                    | (self.embedding_df["id"] == int(response[0][neighbor_index].id))
+                ].text.values[0]
+                + " \n"
+            )
+
+        return context
+
+    def ask_qna(self, question: str) -> str:
+        """Retrieves relevant context using vector search and generates an answer using a QnA model.
+        Args:
+            question (str): The user's question.
+
+        Returns:
+            str: The generated answer from the QnA model, or None if no valid answer could be determined.
+        """
+
+        # Read context from relavent documents
+        self.logger.info("QnA: question: %s", question)
+
+        context = self.find_relevant_context(question)
+        # self.logger.info("QnA: context: %s", context)
+
+        # Get response
+        if len(context) > 0:
+            response = self.model.generate_content(
+                self.prompt_template.format(context=context, question=question)
+            )
+
+            if response and int(response.candidates[0].finish_reason) == 1:
+                answer = response.text
+                self.logger.info("QnA: response: %s", answer)
+                if (
+                    "I cannot determine the answer to that." in answer
+                    or "I could not find any references that are directly related to your question."
+                    in answer
+                ):
+                    return ""
+                else:
+                    return answer
+        return ""
```

Comment:
Try reducing the nesting similar to how I did before. Could make this easier to read.

Reply:
updated the code block.